src/server/telemetry/urlHelpers.js

```diff
@@ -0,0 +1,80 @@
+var url = require('url');
+var { URL } = require('url');
+var merge = require('../../merge');
+
+// This function replicates the relevant logic in node/lib/http.js as closely
+// as possible in order to produce the same result. Therefore, the code is
+// replicated as is, favoring the closest match to the original code without style changes.
+//
+// The code here is only used to build telemetry metadata and is not used to
+// build actual http requests.
+function mergeOptions(input, options, cb) {
+  if (typeof input === 'string') {
+    const urlStr = input;
+    input = urlToHttpOptions(new URL(urlStr));
+  } else if (input && input[url.searchParamsSymbol] &&
+             input[url.searchParamsSymbol][url.searchParamsSymbol]) {
+    // url.URL instance
+    input = urlToHttpOptions(input);
+  } else {
+    cb = options;
+    options = input;
+    input = null;
+  }
+
+  if (typeof options === 'function') {
+    cb = options;
+    options = input || {};
+  } else {
+    options = merge(input || {}, options);
+  }
+  return {options: options, cb: cb};
+}
+
+// This function replicates the relevant logic in node/lib/url.js as closely
+// as possible in order to produce the same result. Therefore, the code is
+// replicated as is, favoring the closest match to the original code without style changes.
+//
+// The code here is only used to build telemetry metadata and is not used to
+// build actual http requests.
+function urlToHttpOptions(url) {
+  const options = {
+    protocol: url.protocol,
+    hostname: typeof url.hostname === 'string' &&
+              url.hostname.startsWith('[') ?
+      url.hostname.slice(1, -1) :
+      url.hostname,
+    hash: url.hash,
+    search: url.search,
+    pathname: url.pathname,
+    path: `${url.pathname || ''}${url.search || ''}`,
+    href: url.href
+  };
+  if (url.port !== '') {
+    options.port = Number(url.port);
+  }
+  if (url.username || url.password) {
+    options.auth = `${url.username}:${url.password}`;
+  }
+  return options;
+}
+
+function constructUrl(options) {
+  var url = options.protocol || 'http:';
+  url += '//';
+  if (options.auth) {
+    url += `${options.auth}@`;
+  }
+  url += options.hostname || options.host || 'localhost';
+  if (options.port) {
+    url += `:${options.port}`
+  }
+  url += options.path || '/';
+
+  return url;
+}
+
+module.exports = {
+  mergeOptions,
+  constructUrl
+};
```

Comment:
I worry these will desynch. If a vulnerability's discovered upstream, is there a way we can reliably receive and act on CVE?

Reply:
I think they are guaranteed to diverge. The goal though is to produce an options object that matches as well as possible the options that will be generated by Node. What is a better way to accomplish that? (Speaking specifically about `mergeOptions` and `urlToHttpOptions`. `constructUrl` is original code, and therefore not relevant to the topic.)

Reply:
Ideally, I suppose, would be rewriting our logic to make the need for these disappear. I don't think that's possible in this case, though. A comment pointing to the original source whence it came would at least allow future maintainers to manually cross-check if changes had been made.

Reply:
The comments do have that. Reviewers of previous Rollbar PRs have asked that comments not contain github URLs or github specific paths, so the references here just show the relative path within the stated project.